### ISSUE TITLE
Feature proposal: separate routing configuration into its own file with an easier-to-use format

### ISSUE THREAD

**Issue by xkxx** [v2ray/v2ray-core#422]
**Version**: 2.20.2 (latest)
**Usage scenario**: ad-blocking

*(This issue is related to https://github.com/v2ray/v2ray-core/issues/329)*

Hi all,

I'd like to propose an alternative easier-to-use format for routing configuration. As the number of rules grow over time to address needs for [ad-blocking](https://www.v2ray.com/chapter_05/00_config/09_ad_blocker.html) and [gfwlist-based whitelisting](https://github.com/v2ray/v2ray-core/issues/329), the json-based config file becomes very unwieldy and hard to maintain. It might be a good idea to move these routing rules into its own file. It could use a csv format similar to the hosts file, something like this:

```csv
# format: outboundTag  rule_type  value  

# routing based on IP
direct  ip  0.0.0.0/8
direct  ip  10.0.0.0/8

# block based on domain name
adblock  domain  msg.71.am
adblock  domain  17un.co

# leave value blank if none needed
direct  chinasites  
```

It has the following advantages:

* Separating routing config from the rest of v2ray configuration, making it easier to load and modify the rest of the config file in a text editor.
* Providing a much less cluttered and more human readable format for routing. csv / hosts file is easier on the eyes than json and less error-prone to edit as well.
* Better suited for shell scripting. echo / pipe / append all work on the csv format, unlike json.
* Non-binary compared to protobuf, making it human-readable and editable.

To support this new format, these changes are needed on v2ray-core:

* Addition of a new CLI option, `-routes <file-name>`
* Internally, the new routing file will be translated into protobuf format and fed to the existing router module, so no extensive changes to router is needed.

 Comments / thoughts? If interested, I can prepare a PR for this in the coming weeks. Thanks!

**Comment by xiaokangwang** [comment]
The support for plugin in V2Ray in the road map and will be implemented in the near future.

Once plugin support is complete, it will be possible to support arbitrary format without modifying v2ray-core code base.

Alternatively, it isn't hard to write a tool that convert the format you proposed into conventional json format and merge it into your configure file. In this way, we will able to keep v2ray-core focus on it's primary object and decease code needed to be maintained as well as attack vectors. 

By the way, as far as I know, there is a major non-pushed modification authored by the primary developer, it could be hard to rebase if you decide to fork from the current head.

The idea above is for informational purpose only, @DarienRaymond will decide whether your code can be merged into origin.

**Comment by xkxx** [comment]
hey @xiaokangwang,

Thanks for the excellent feedback! I like the idea of writing a tool for merging config files and I'll be doing that for sure as a stopgap solution before plugin support is landed. I do think something like my original proposal in core would be very beneficial to a wide range of users looking to use v2ray for beyond anti-censorship though. Regardless, I'd love to hear what others think of this idea!

**Comment by DarienRaymond** [comment]
As you already know, the core config is now purely based on protobuf. The [JSON parser](https://github.com/v2ray/v2ray-core/tree/master/tools/conf) is provided only for backward compatibility. In the future, core will not touch files directly. All config parsing logic will be done before core is launched.

If you wish to create a new format of config file, you may do it as a wrapper.  The wrapper reads a config file in your format, translate it into protobuf and feed into V2Ray core. The advantage is you may use any language that you prefer to create the wrapper, as protobuf is a language-independent technology.

**Comment by DarienRaymond** [comment]
Feel free to open another issue at https://github.com/v2ray/planning for further discussion.
